vectorcode.chromadb: enable check phase

MattSturgeon opened the pull request:
Follow up to https://github.com/NixOS/nixpkgs/pull/417720#discussion_r2153550800

I've marked @sarahec as the author of 3cf4c4b9dd483dfbc72360bd69281dcd4e5a3dfb, as it is lifted directly from their work in #412528. Note most of the changes in #421528 aren't needed because here we're overriding the v1.0 base package, and most things stay the same.

Additionally, ~~the `broken` flag has been restored to chromadb on aarch64-linux.~~ This was dropped in https://github.com/NixOS/nixpkgs/pull/412528, but the brokenness was just being _hidden_ by disabling the import check on aarch64-linux. The package is _still_ broken due to an onnxruntime issue (https://github.com/microsoft/onnxruntime/issues/10038). EDIT: this issue only occurs in the build sandbox, so I haven't marked it as broken.

cc @fabaff @sarahec @dotlambda @GaetanLepage

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

- Added `vectorcode.chromadb` and `vectorcode.python` passthrus to expose overridden inputs and aid debugging
- Removed `doCheck = false`
- Added disabled test list from #412528
- ~~Restored chromadb's `broken` flag on aarch64-linux~~
- ~~Removed condition from chromadb's `pythonImportsCheck`~~

---

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

sarahec committed:
vectorcode.chromadb: enable check phase

Using a subset of disabled tests from the base package, as others only
apply to >1.0

---

Additionally, disable test_chroma.py and test_client.py:

These tests have issues running in parallel, such as when building the
python 3.12 and 3.13 versions simultaneously.

E.g. when running nixpkgs-review or unlucky timing in hydra.

Usually issues show up as:
    ValueError: An instance of Chroma already exists for ephemeral with different settings

MattSturgeon committed:
vectorcode: add python passthru

This allows accessing the overridden chromadb in the overridden python
instance.

sarahec commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 417811`
Commit: `abb367b29d9c260ee51603f3bc42a0baf712e11e`

---
### `aarch64-darwin`
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vimPlugins.vectorcode-nvim</li>
  </ul>
</details>

MattSturgeon commented:
I assume it is better to disable the entire [`test_chroma.py` file](https://github.com/chroma-core/chroma/blob/0.6.3/chromadb/test/test_chroma.py), rather than just the `test_local_db` test, as that file is mostly all database tests that I assume have concurrency issues?

Also, I assume we want to disable on `buildPlatform.isDarwin` rather than `hostPlatform.isDarwin`? Since the concurrency issues relate to the actual builder platform not the final runtime target platform.

sarahec commented:
> > Use `nixpkgs-review --build-args "\-j1"`.
> 
> Is this an issue for hydra too, or just nixpkgs-review?
> 
> I'd prefer to have as many tests enabled as possible, but if this will cause disruption for the package maintainers then disabling is fine too.
> 
> If it can cause problems on hydra, then it should be disabled; no question.

It can cause problems on Hydra as well, depending on timing. Better to disable the file.

dotlambda commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 417811`
Commit: `621dfce25a1346de26679601e7104b3fd50a32ea`

---
### `x86_64-linux`
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vimPlugins.vectorcode-nvim</li>
  </ul>
</details>

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 417811`

Logs: https://github.com/MattSturgeon/nixpkgs-review-gha/actions/runs/15767710932


---
### `x86_64-linux` (sandbox = true)
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vimPlugins.vectorcode-nvim</li>
  </ul>
</details>


---
### `aarch64-linux` (sandbox = true)
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vimPlugins.vectorcode-nvim</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-linux`</summary>
<details>
<summary>vectorcode</summary>
<pre>  File "/nix/store/ldnnbghlrg7zm8cgwx20nd3yvvm1c3vp-python3.13-pytest-8.3.5/lib/python3.13/site-packages/_pytest/main.py", line 975 in genitems
  File "/nix/store/ldnnbghlrg7zm8cgwx20nd3yvvm1c3vp-python3.13-pytest-8.3.5/lib/python3.13/site-packages/_pytest/main.py", line 809 in perform_collect
  File "/nix/store/ldnnbghlrg7zm8cgwx20nd3yvvm1c3vp-python3.13-pytest-8.3.5/lib/python3.13/site-packages/_pytest/main.py", line 347 in pytest_collection
  File "/nix/store/wgcd872b1163l1l1r16f58wkmjhzw76s-python3.13-pluggy-1.6.0/lib/python3.13/site-packages/pluggy/_callers.py", line 121 in _multicall
  File "/nix/store/wgcd872b1163l1l1r16f58wkmjhzw76s-python3.13-pluggy-1.6.0/lib/python3.13/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/nix/store/wgcd872b1163l1l1r16f58wkmjhzw76s-python3.13-pluggy-1.6.0/lib/python3.13/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/nix/store/ldnnbghlrg7zm8cgwx20nd3yvvm1c3vp-python3.13-pytest-8.3.5/lib/python3.13/site-packages/_pytest/main.py", line 336 in _main
  File "/nix/store/ldnnbghlrg7zm8cgwx20nd3yvvm1c3vp-python3.13-pytest-8.3.5/lib/python3.13/site-packages/_pytest/main.py", line 283 in wrap_session
  File "/nix/store/ldnnbghlrg7zm8cgwx20nd3yvvm1c3vp-python3.13-pytest-8.3.5/lib/python3.13/site-packages/_pytest/main.py", line 330 in pytest_cmdline_main
  File "/nix/store/wgcd872b1163l1l1r16f58wkmjhzw76s-python3.13-pluggy-1.6.0/lib/python3.13/site-packages/pluggy/_callers.py", line 121 in _multicall
  File "/nix/store/wgcd872b1163l1l1r16f58wkmjhzw76s-python3.13-pluggy-1.6.0/lib/python3.13/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/nix/store/wgcd872b1163l1l1r16f58wkmjhzw76s-python3.13-pluggy-1.6.0/lib/python3.13/site-packages/pluggy/_hooks.py", line 512 in __call__
  File "/nix/store/ldnnbghlrg7zm8cgwx20nd3yvvm1c3vp-python3.13-pytest-8.3.5/lib/python3.13/site-packages/_pytest/config/__init__.py", line 175 in main
  File "/nix/store/ldnnbghlrg7zm8cgwx20nd3yvvm1c3vp-python3.13-pytest-8.3.5/lib/python3.13/site-packages/_pytest/config/__init__.py", line 201 in console_main
  File "/nix/store/ldnnbghlrg7zm8cgwx20nd3yvvm1c3vp-python3.13-pytest-8.3.5/lib/python3.13/site-packages/pytest/__main__.py", line 9 in <module>
  File "<frozen runpy>", line 88 in _run_code
  File "<frozen runpy>", line 198 in _run_module_as_main

Extension modules: numpy._core._multiarray_umath, numpy.linalg._umath_linalg (total: 2)
/nix/store/2r1kmw6cqp7y5irycz5sxha20qs0i32v-pytest-check-hook/nix-support/setup-hook: line 23:   492 Aborted                 (core dumped) /nix/store/avgafrfq80hx92p98m60wh2khmnmxrzc-python3-3.13.4/bin/python3.13 "${flagsArray[@]}"</pre>
</details>
</details>


---
### `x86_64-darwin` (sandbox = true)
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vimPlugins.vectorcode-nvim</li>
  </ul>
</details>


---
### `aarch64-darwin` (sandbox = true)
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vimPlugins.vectorcode-nvim</li>
  </ul>
</details>

dotlambda commented on pkgs/by-name/ve/vectorcode/package.nix:
```diff
@@ -8,6 +8,7 @@
   python3,
   rustc,
   rustPlatform,
+  stdenv,
```
```suggestion
```

MattSturgeon commented:
> **`aarch64-linux` (sandbox = true)**
> ❌ 3 packages failed to build:
> * vectorcode
> * vectorcode.dist
> * vimPlugins.vectorcode-nvim


The error in the logs doesn't seem related the concurrency issue, which seems to have been fixed by `pytestFlags = [ "--numprocesses=1" ]`.

When introduced by @GaetanLepage in https://github.com/NixOS/nixpkgs/pull/394565, vectorcode was marked "broken" on aarch64-linux. This seems to have been inherited from chromadb.

It looks like https://github.com/NixOS/nixpkgs/pull/412528 (accidentally?) removed `meta.broken` from chromadb in [`e54888cf`](https://github.com/NixOS/nixpkgs/commit/e54888cf51f66852e94a0a29df2ae6edc0a625bb#diff-76aaaa9544d4848d5d01c649fbc0a2881b25f988ab7ac57525d0b379b92c2e30L180), so I'll add it back in a separate commit here.

---

Should `--numprocesses=1` and/or `meta.broken = isLinux && isAarch64` be added to the base package or the overridden package?

MattSturgeon commented on pkgs/by-name/ve/vectorcode/package.nix:
```diff
@@ -45,7 +46,51 @@ let
         ];
-        doCheck = false;
+
+        pytestFlags = [
+          # This package sometimes has issues running multiple tests in parallel.
+          # ValueError: An instance of Chroma already exists for ephemeral with different settings
+          "--numprocesses=1"
+        ];
```
Is this relevant to the base package too? Should I move this to [pkgs/development/python-modules/chromadb/default.nix](https://github.com/NixOS/nixpkgs/blob/9541941f0405c4d1145923a436b205458d0b4d2a/pkgs/development/python-modules/chromadb/default.nix) ?

sarahec replied:
Please remove this and disable the affected test instead. No other package in nixpkgs resorts to this flag.

MattSturgeon commented on pkgs/by-name/ve/vectorcode/package.nix:
```diff
@@ -45,7 +46,51 @@ let
+        ];
+
+        meta = old.meta // {
+          broken = stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isAarch64;
```
Same question: Should I move this to [pkgs/development/python-modules/chromadb/default.nix](https://github.com/NixOS/nixpkgs/blob/9541941f0405c4d1145923a436b205458d0b4d2a/pkgs/development/python-modules/chromadb/default.nix) ?

MattSturgeon commented:
> Test collection is breaking on aarch64-Linux, probably due to a missing package that's used in the tests. That doesn't mean the package is broken.

I'll try to look into that, when I get time. It may end up being out of scope for this PR.

> ~~In fact, we had it building and running fine (via the vectorcode tests) already.~~ I can't confirm that. Could you run nixpkgs-review on the closed #417720? The outcome of that should determine whether to mark this as broken.

I tried running nixpkgs-review on it, but ran into https://github.com/Mic92/nixpkgs-review/issues/439 ([logs here](https://github.com/MattSturgeon/nixpkgs-review-gha/actions/runs/15773727284)).

I've not setup emulation yet and I don't have access to any remote builders, so I can't test ARM Linux locally.

> You may want to peruse the diff in https://github.com/NixOS/nixpkgs/pull/412528/commits/61265017a39b23227a8502fbfcdfce6b509f18e7 again in case something was missed.

I'll take another look when I get chance, but from a quick skim I can't see many inputs being _removed_ in that diff.

---

Thanks for steering me in the right direction.

MattSturgeon commented:
> > Test collection is breaking on aarch64-Linux, probably due to a missing package that's used in the tests. That doesn't mean the package is broken.
> 
> I'll try to look into that, when I get time. It may end up being out of scope for this PR.

Looking at [the logs](https://github.com/MattSturgeon/nixpkgs-review-gha/actions/runs/15767710932/job/44447909460#step:6:4287) in more detail, it looks like chromadb built fine, but vectorcode is failing to import it during _its_ check phase.

```
vectorcode> Executing pytestCheckPhase
vectorcode> pytest flags: -m pytest -k not\ \(test_get_embedding_function\)\ and\ not\ \(test_get_embedding_function_fallback\)\ and\ not\ \(test_get_reranker\)\ and\ not\ \(test_supported_rerankers_initialization\)
vectorcode> ============================= test session starts ==============================
vectorcode> platform linux -- Python 3.13.4, pytest-8.3.5, pluggy-1.6.0
vectorcode> rootdir: /build/source
vectorcode> configfile: pyproject.toml
vectorcode> plugins: anyio-4.9.0, typeguard-4.4.2, hypothesis-6.131.17
vectorcode> collecting ... Fatal Python error: Aborted
vectorcode> 
vectorcode> Current thread 0x0000fffff7ff5ec0 (most recent call first):
vectorcode>   File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
vectorcode>   File "<frozen importlib._bootstrap_external>", line 1320 in create_module
vectorcode>   File "<frozen importlib._bootstrap>", line 813 in module_from_spec
vectorcode>   File "<frozen importlib._bootstrap>", line 921 in _load_unlocked
vectorcode>   File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
vectorcode>   File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
vectorcode>   File "/nix/store/x2vh7cbf73sbaxp7xzadcfz9fiq5pj7d-python3.13-onnxruntime-1.22.0/lib/python3.13/site-packages/onnxruntime/capi/_pybind_state.py", line 32 in <module>
vectorcode>   File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
vectorcode>   File "<frozen importlib._bootstrap_external>", line 1026 in exec_module
vectorcode>   File "<frozen importlib._bootstrap>", line 935 in _load_unlocked
vectorcode>   File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
vectorcode>   File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
vectorcode>   File "/nix/store/x2vh7cbf73sbaxp7xzadcfz9fiq5pj7d-python3.13-onnxruntime-1.22.0/lib/python3.13/site-packages/onnxruntime/__init__.py", line 24 in <module>
vectorcode>   File "<frozen importlib._bootstrap>", line 488 in _call_with_frames_removed
vectorcode>   File "<frozen importlib._bootstrap_external>", line 1026 in exec_module
vectorcode>   File "<frozen importlib._bootstrap>", line 935 in _load_unlocked
vectorcode>   File "<frozen importlib._bootstrap>", line 1331 in _find_and_load_unlocked
vectorcode>   File "<frozen importlib._bootstrap>", line 1360 in _find_and_load
vectorcode>   File "<frozen importlib._bootstrap>", line 1387 in _gcd_import
vectorcode>   File "/nix/store/avgafrfq80hx92p98m60wh2khmnmxrzc-python3-3.13.4/lib/python3.13/importlib/__init__.py", line 88 in import_module
vectorcode>   File "/nix/store/ppcaa9i4n81z5w35b671y60jhlfnjsvj-python3.13-chromadb-0.6.3/lib/python3.13/site-packages/chromadb/utils/embedding_functions/onnx_mini_lm_l6_v2.py", line 63 in __init__
vectorcode>   File "/nix/store/ppcaa9i4n81z5w35b671y60jhlfnjsvj-python3.13-chromadb-0.6.3/lib/python3.13/site-packages/chromadb/utils/embedding_functions/__init__.py", line 57 in DefaultEmbeddingFunction
vectorcode>   File "/nix/store/ppcaa9i4n81z5w35b671y60jhlfnjsvj-python3.13-chromadb-0.6.3/lib/python3.13/site-packages/chromadb/api/models/CollectionCommon.py", line 112 in CollectionCommon
vectorcode>   File "/nix/store/ppcaa9i4n81z5w35b671y60jhlfnjsvj-python3.13-chromadb-0.6.3/lib/python3.13/site-packages/chromadb/api/models/CollectionCommon.py", line 100 in <module>
```

From the chromadb part of the trace:

[`chromadb/utils/embedding_functions/onnx_mini_lm_l6_v2.py, line 63 in __init__`](https://github.com/chroma-core/chroma/blob/0.6.3/chromadb/utils/embedding_functions/onnx_mini_lm_l6_v2.py#L63)
```py
try:
    # Equivalent to import onnxruntime
    self.ort = importlib.import_module("onnxruntime") # <-- here
except ImportError:
    raise ValueError(
        "The onnxruntime python package is not installed. Please install it with `pip install onnxruntime`"
    )
```

Could this be an ARM-specific bug in [`importlib.import_module`](https://docs.python.org/3/library/importlib.html#importlib.import_module), meaning a `Fatal Python error` occurs instead of throwing the expected [`ImportError`](https://docs.python.org/3/library/exceptions.html#ImportError)?

MattSturgeon commented:
Ah, I see.

The base chromadb package has: https://github.com/NixOS/nixpkgs/blob/c42d484a470999f8bcfd322d34168efd3ad5f2f4/pkgs/development/python-modules/chromadb/default.nix#L172-L174

However IIUC that actually means the package _is_ broken on ARM linux, it is just hiding it by skipping the import check.

If so, we should do:

```diff
diff --git a/pkgs/development/python-modules/chromadb/default.nix b/pkgs/development/python-modules/chromadb/default.nix
index 868cd5a7ca42..704df33ccd31 100644
--- a/pkgs/development/python-modules/chromadb/default.nix
+++ b/pkgs/development/python-modules/chromadb/default.nix
@@ -169,9 +169,7 @@ buildPythonPackage rec {
     starlette
   ];
 
-  # Disable on aarch64-linux due to broken onnxruntime
-  # https://github.com/microsoft/onnxruntime/issues/10038
-  pythonImportsCheck = lib.optionals (stdenv.hostPlatform.system != "aarch64-linux") [ "chromadb" ];
+  pythonImportsCheck = [ "chromadb" ];
 
   # Test collection breaks on aarch64-linux
   doCheck = stdenv.hostPlatform.system != "aarch64-linux";
@@ -271,5 +269,8 @@ buildPythonPackage rec {
       sarahec
     ];
     mainProgram = "chroma";
+    # Broken on aarch64-linux due to onnxruntime issue:
+    # https://github.com/microsoft/onnxruntime/issues/10038
+    broken = with stdenv.hostPlatform; isAarch && isLinux;
   };
 }
```

MattSturgeon committed:
python3Packages.chromadb: improve doCheck condition

On aarch64-linux onnxruntime cannot be imported, as it attempts to read
/sys/devices/system/cpu, which does not exist in the sandbox.

As onnxruntime cannot be imported, chromadb cannot be tested.

However, this issue only applies in the build sandbox, not at runtime.
Therefore the package doesn't need to be marked broken.

Updated the condition to check `buildPlatform` instead of `hostPlatform`,
as the issue relates to the build sandbox and is not a runtime issue.

Use `isAarch` and `isLinux` attributes instead of doing string comparison.

---

See https://github.com/NixOS/nixpkgs/pull/412528#discussion_r2129396116
And upstream onnxruntime issue: https://github.com/microsoft/onnxruntime/issues/10038

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 417811`

Logs: https://github.com/MattSturgeon/nixpkgs-review-gha/actions/runs/15776610565


---
### `x86_64-linux` (sandbox = true)
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vimPlugins.vectorcode-nvim</li>
  </ul>
</details>



---
### `x86_64-darwin` (sandbox = true)
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vimPlugins.vectorcode-nvim</li>
  </ul>
</details>


---
### `aarch64-darwin` (sandbox = true)
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vimPlugins.vectorcode-nvim</li>
  </ul>
</details>

MattSturgeon committed:
vectorcode: inherit `doCheck` from chromadb

The tests need to import onnxruntime (a transitive of chromadb),
however it has issues running in the build sandbox on aarch64-linux.

It may work at runtime, so the package is not marked broken, however it
will break tests, so the check phase is disabled when building on
aarch64-linux.

Since the issue comes via chromadb, we inherit its `doCheck` attribute.

MattSturgeon committed:
vectorcode: move installShellFiles to nativeBuildInputs

It was accidentally in `nativeCheckInputs`, so the build fails when
`doCheck = false`.

MattSturgeon commented:
> Please follow commit message guidelines for the chromadb commit.

Oops, I forgot it was in the `python3Packages` package set. Will fix next push. It would've been helpful if you'd said explicitly what was wrong with the commit message; that would've saved me a few minutes of confusion.

> ❌ 3 packages failed to build:
> * vectorcode
> * vectorcode.dist
> * vimPlugins.vectorcode-nvim

Looks like we have the same concurrency issues testing `test_client.py` too:

```
 python3.13-chromadb> =========================== short test summary info ============================
python3.13-chromadb> ERROR chromadb/test/test_client.py::test_ephemeral_client - ValueError: An instance of Chroma already exists for ephemeral with different settings
python3.13-chromadb> !!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
python3.13-chromadb> !!!!!!!!!!!! xdist.dsession.Interrupted: stopping after 1 failures !!!!!!!!!!!!!
python3.13-chromadb> = 221 passed, 12 skipped, 1 xfailed, 108873 warnings, 1 error in 85.79s (0:01:25) =
```

https://github.com/MattSturgeon/nixpkgs-review-gha/actions/runs/15779335576/job/44480884102#step:6:6881

---

In fact, this makes me question:

> The problem is that a fast enough machine winds up building python 3.12 and python 3.13 versions simultaneously.

Because in this build, _only_ the python 3.13 version is being built; there is no python 3.12 dependency for vectorcode.

This makes me suspect that `--numprocesses=1` should work, at least for `vectorcode.chromadb`. I'm curious what @sarahec was building when they [reproduced](https://github.com/NixOS/nixpkgs/pull/417811#issuecomment-2989637766) the concurrency issue _with_ `--numprocesses=1`?

sarahec commented on pkgs/by-name/ve/vectorcode/package.nix:
```diff
@@ -45,7 +45,46 @@ let
+          # Hypothesis FailedHealthCheck due to nested @given tests
+          "chromadb/test/cache/test_cache.py"
+
+          # Tests fail when running in parallel.
+          # E.g. when building the building python 3.12 and 3.13 versions simultaneously.
+          # ValueError: An instance of Chroma already exists for ephemeral with different settings
+          "chromadb/test/test_chroma.py"
```
We've never seen the concurrency issue on the base package. It has to do with the way the tests were set up (and was fixed between 0.6.3 and now).

sarahec replied:
P.S. Technically, these last two tests could be excluded on Darwin only.

MattSturgeon replied:
> P.S. Technically, these last two tests could be excluded on Darwin only. 

I reproduced the issue on x86_64-linux here: https://github.com/NixOS/nixpkgs/pull/417811#issuecomment-2987864344

MattSturgeon commented on pkgs/development/python-modules/chromadb/default.nix:
```diff
@@ -171,10 +171,10 @@ buildPythonPackage rec {
 
   # Test collection breaks on aarch64-linux
-  doCheck = stdenv.hostPlatform.system != "aarch64-linux";
+  doCheck = with stdenv.buildPlatform; !(isAarch && isLinux);
```
That's incorrect.

`buildPlatform` is the platform that we're building on; i.e. where the build script is running, `hostPlatform` is the platform that we're building for; i.e. where users will install/run the package.

You're right, you _can_ detect cross-comp using `lib.systems.equals buildPlatform hostPlatform`,[^1][^2] but that's not its only use.

You should always think about whether you actually care about the build platform of the host platform.

[^1]: Most of the time you actually want to check `buildPlatform.canExecute`. hostPlatform` instead of checking equality 
[^2]: A lot of legacy code uses `==` and `!=` to compare platforms, but this is a bad idea, and will likely get deprecated.

MattSturgeon replied:
_Specifically_ for `doCheck`, the distinction shouldn't matter; IIRC the check phase only runs when `buildPlatform.canExecute hostPlatform`. That's because it wouldn't make sense to test host exes on the build platform, if it can't execute them.

Although I could be misremembering, and it still feels like good practice to use the "correct" platform for the intention regardless.

sarahec replied:
TIL. Thank you.

sarahec commented:
Apologies for being short with people the last few days. Being sick apparently gives me a short temper. (Still recovering, but no longer coughing up a lung.)

MattSturgeon commented:
> > I get your argument, that the `aarch64-linux` package works when used in a normal runtime environment. But it seems problematic that _any_ package that tries to import chromadb will fail while in a build sandbox.
> 
> [`nixpkgs-review`](https://github.com/NixOS/nixpkgs/pull/412528#issuecomment-2953163606) built the 33 packages that import chromadb on aarch64-linux (sandbox==true).

Interesting. So those packages must each either not invoke whichever part of chromadb imports onnxruntime during their tests, or disable any tests that do on aarch64-linux.

sarahec commented:
re: python 3.12 and python 3.13 building concurrently: it's a known issue on Darwin, but I guess its not the issue here after all. I have a theory and I'll see if I can get `nixpkgs-review` to emit a build graph.

sarahec commented:
> > [`nixpkgs-review`](https://github.com/NixOS/nixpkgs/pull/412528#issuecomment-2953163606) built the 33 packages that import chromadb on aarch64-linux (sandbox==true).
> 
> Interesting. So those packages must each either not invoke whichever part of chromadb imports onnxruntime during their tests, or disable any tests that do on aarch64-linux.

As far as I can tell, this is happening only when 1) `onnxruntime` is called to detect the environment 2) and when sandboxed. By the time a package imports `chromadb`, stage 1 has passed -- it's getting the binary from the build cache (or, if rebuilding from source, disabling the imports check handles this case).

It was tempting to simply mark this as broken on aarch64-linux, but then I try to put myself in the user's shoes and go the extra mile to make it work (just as you did with tests ;) ).

sarahec commented:
@GaetanLepage please take a look

GaetanLepage reviewed (approved):
Thanks!

GaetanLepage commented:
Thanks @MattSturgeon for the very clean commit history with precise descriptions!
